### PR TITLE
[PORT] - HD Wallet phase 2

### PR DESCRIPTION
**Pull request by ptschip**
Still a work in progress.  Adding several other followup commits, some cleanup , some minor new rpc features.

### CODE REVIEW

**Comment by ptschip** [comment]
Phase 2 of HD wallet is complete...

File: src/rpc/misc.cpp [hunk 1]
```diff
@@ -185,6 +185,8 @@ UniValue validateaddress(const UniValue &params, bool fHelp)
             "  \"iscompressed\" : true|false,  (boolean) If the address is compressed\n"
             "  \"account\" : \"account\"         (string) DEPRECATED. The account associated with the address, \"\" is "
             "the default account\n"
+            "  \"hdkeypath\" : \"keypath\"       (string, optional) The HD keypath if the key is HD and available\n"
```
Comment (gandrewstone) [inline src/rpc/misc.cpp:188]: "The HD keypath" ... the help should not repeat the words in the key.  Does this refer to https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#the-key-tree?  Perhaps: The path required to derive this key given the key tree root (that is, this key's location in the tree, e.g. "m/2/6/3/...")

File: src/rpc/misc.cpp [hunk 2]
```diff
@@ -185,6 +185,8 @@ UniValue validateaddress(const UniValue &params, bool fHelp)
...
+            "  \"hdmasterkeyid\" : \"<hash160>\" (string, optional) The Hash160 of the HD master pubkey\n"
```
Comment (gandrewstone) [inline src/rpc/misc.cpp]: (hex string, optional)

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -2531,6 +2531,7 @@ UniValue getwalletinfo(const UniValue &params, bool fHelp)
                             "Jan 1 1970 GMT) that the wallet is unlocked for transfers, or 0 if the wallet is locked\n"
                             "  \"paytxfee\": x.xxxx,         (numeric) the transaction fee configuration, set in " +
             CURRENCY_UNIT + "/kB\n"
+                            "  \"hdmasterkeyid\": \"<hash160>\", (string) the Hash160 of the hd master pubkey\n"
```
Comment (gandrewstone) [inline src/wallet/rpcwallet.cpp]: (hex string)

File: src/rpc/misc.cpp (hunk 2 above)
Comment (ptschip) [reply src/rpc/misc.cpp]: done

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (ptschip) [reply src/wallet/rpcwallet.cpp]: done

File: src/rpc/misc.cpp (hunk 1 above)
Comment (ptschip) [reply src/rpc/misc.cpp:188]: no this has nothing to do with the master private key or any secret way to derive the master key. This is just the key derivation path.  How the keys are derived from the master key.  As you mentioned, "m/2/6/3/" etc...

**Pull request merged**